crates/ruff/tests/lint.rs

```diff
@@ -1248,3 +1248,39 @@ fn negated_per_file_ignores_absolute() -> Result<()> {
     });
     Ok(())
 }
+
+/// patterns are additive, can't use negative patterns to "un-ignore"
```

Comment:
Is this something we are locked into? I can imagine this being somewhat unintuitive since things like `.gitignore` lets you un-ignore things.

Reply:
I'm not set on this behavior, just was favoring the simplest option to understand, unless we had a clear need for something more complex. But it's true that it will be hard to change this once shipped, so it's good to think through what we prefer now.

The main difference I see between this case and gitignore is that a negative pattern in gitignore serves exactly one purpose: to un-ignore whatever it matches. It doesn't also mean "ignore everything that doesn't match this pattern."

In our case we do want "ignore everything that doesn't match this pattern" for negative patterns, so there's more risk of overloading the meaning if we also add "un-ignore everything that does match this pattern."

For instance, here's a case to consider:

```
# ignore D rules in all __init__.py files
"__init__.py" = ["D"]
# ignore D rules everywhere outside src/
# - does this also mean "actually, don't ignore in __init__.py files inside src/?
"!src/**.py" = ["D"]
```

I'm not sure which behavior is more intuitive for this example. But if we did add un-ignoring, and you did want to ignore in all `__init__.py` files inside `src/` too, you can always just switch the order of the two patterns; the rule would be "last one wins." I think between this and the option of adding another more-specific positive pattern after the negative one, there shouldn't be any cases you can't express at all, if we add this.

Initially it seemed odd to me that a negative pattern would un-ignore on failure to hit, while a positive pattern wouldn't do that, but after thinking through some examples, I think that behavior works out OK; for a negative pattern the no-hit cases are more narrowly targeted, so it's more intuitive to un-ignore in that case.

I guess my last concern: is it too confusing if a negative pattern in `extend-per-file-ignores` (in a different config location from `per-file-ignores`) overrides some of what's specified in `per-file-ignores`? Or is this fine?

cc @charliermarsh @zanieb for opinions also

Reply:
The typical way I like to do filtering for whitelist/blacklist situations like this is as follows:

* Every rule is either a whitelist or a blacklist rule.
* The rules are ordered. Later rules take priority over earlier rules.
* When there are 1 or more whitelist rules, then in order for a subject to pass the filter, the last matching rule must be a whitelist rule.
* When there are 0 whitelist rules, a subject passes the filter when it matches 0 blacklist rules.
* Otherwise, the filter rejects the subject.

I find that these semantics overall provide a lot of flexibility. But you're right that gitignore doesn't neatly map on to this model. The other complication that seems present here is that the rules aren't just ignore/not-ignore, but they also come with lint rules associated with the filter. I haven't thought through fully how that might impact things.

I think the model above overall matches what you're saying around adding un-ignoring, particularly with the last rule matching being the winner. I think I overall personally like this style. I don't have a good answer for what to do about `extend-per-file-ignores` though.

Reply:
After more discussion off-line, it feels like allowing un-ignores is too unpredictable. The case that seems like the most obvious for un-ignores is something like this:

```
[lint.per-file-ignores]
"src/**/*.py" = "[X]"
"!src/foo/bar/**/*.py" = "[X]"
```

But even that is subtly wrong, because it doesn't just make a carve-out from the prior rule, it ends up ignoring `X` _everywhere_ outside `src/foo/bar`, rendering the first rule redundant.

I'm going to roll back this PR to just test the existing purely-additive behavior (which we actually already released this morning); my suspicion is that if we want per-file selects it will be a lot clearer to add a `per-file-selects` config. But we should have more discussion and do that in a separate PR.

Reply:
SGTM. Thanks for thinking through this. :)